---
#MIT License
#
#Copyright (c) 2023 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "del_params_dsns"
output: html_document
date: '2022-12-07'
---

```{r}

library(data.table)
library(ggpubr)
library(ggplot2)
cbbPalette <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


```


```{r}

continuous_variables <- c(
  'gc_content',
  'inculture',
  'inplanta'
)

legend_labels_continuous <- c(
  'GC Content',
  'Normalized in Culture RNAseq Signal',
  'Normalized in Planta RNAseq Signal'
)

axis_xlim_lowers <-c(
  0.25,
  0,
  0
)

axis_xlim_uppers <-c(
  0.75,
  300,
  300
)

```


```{r}

for (i in seq(from=1,to=length(continuous_variables))){
  
  variable = continuous_variables[i]
    
  legend_label = legend_labels_continuous[i]
  
  axis_xlim_lower = axis_xlim_lowers[i]
  
  axis_xlim_upper = axis_xlim_uppers[i]
  
  print(i)
  print(variable)
  print(legend_label)
  
  ## read in values for randomly selected regions
  vector_shuffled_rice <- data.frame(fread(paste('../../pav_newest_gladieux_only_fungap/sv_callers/shuffled.',
                                            variable,'.txt', sep='')))$V1
  
  ## read in values for deletions
  vector_del_rice <- data.frame(fread(paste('../../pav_newest_gladieux_only_fungap/sv_callers/del.',
                                          variable,'.txt', sep='')))$V1
  
  # same for wheat blast
  vector_shuffled_wheat <- data.frame(fread(paste('../../pav_newest_wheat_blast_all/sv_callers/shuffled.',
                                            variable,'.txt', sep='')))$V1
  
  vector_del_wheat <- data.frame(fread(paste('../../pav_newest_wheat_blast_all/sv_callers/del.',
                                          variable,'.txt', sep='')))$V1
  
  plot_df <- data.frame(
    value = c(vector_shuffled_rice, vector_del_rice,
              vector_shuffled_wheat, vector_del_wheat),
    group = c(rep('MoO Baseline',length(vector_shuffled_rice)),
              rep('MoO Indels',length(vector_del_rice)),
              rep('MoT Baseline',length(vector_shuffled_wheat)),
              rep('MoT Indels',length(vector_del_wheat)))
  )
  
  # plot density hists for each variable
  
  p <- ggplot(data=plot_df, aes(x=value, y=group, fill=group))+
    geom_violin(lwd=0.25, scale="width", adjust=5) + 
    geom_boxplot(width = 0.15, outlier.shape = NA,lwd=0.25,coef=0)+
    stat_summary(fun = "mean",
         geom = "point",
         color = "black",
         size=0.5)+
    xlab(legend_label)+
    ylab('')+
    xlim(axis_xlim_lower,axis_xlim_upper)+
    theme_classic()+
    theme(plot.title = element_text(size=8,hjust = 0.5), 
          axis.text=element_text(size=6),axis.title=element_text(size=8) , 
          legend.title=element_text(size=8), legend.position = 'bottom', 
          legend.text=element_text(size=6)) + theme(legend.title = 
                                                      element_blank())+
    theme(legend.position = "none")+
    scale_fill_manual(values=c(cbbPalette[4], "#5f5f5f", cbbPalette[2], cbbPalette[3]))
  
  ggsave(paste(variable, "_", 'density_plot',
               '.pdf',sep = ''), plot = p, width = 6.5/2, height = 7.5/4)

}


```

```{r}

continuous_variables <- c(
  'methylation',
  'H3K36me3',
  'H3K27me3',
  'H3K27ac',
  'eccdna'
)

legend_labels_continuous <- c(
  '% Methylated Cytosines',
  'Normalized H3K36me3 Signal',
  'Normalized H3K27me3 Signal',
  'Normalized H3K27ac Signal',
  'Normalized EccDNA-Seq Signal'
)

axis_xlim_lowers <-c(
  0,
  0,
  0,
  0,
  0
)

axis_xlim_uppers <-c(
  2,
  1500,
  1500,
  1500,
  500
)

```

```{r}

## continuous variables that are only present in rice blast
for (i in seq(from=1,to=length(continuous_variables))){
  
  variable = continuous_variables[i]
    
  legend_label = legend_labels_continuous[i]
  
  axis_xlim_lower = axis_xlim_lowers[i]
  
  axis_xlim_upper = axis_xlim_uppers[i]
  
  print(i)
  print(variable)
  print(legend_label)
  
  ## read in values for shuffled regions
  vector_shuffled_rice <- data.frame(fread(paste('../../pav_newest_gladieux_only_fungap/sv_callers/shuffled.',
                                            variable,'.txt', sep='')))$V1
  
  ## read in values for deletions
  vector_del_rice <- data.frame(fread(paste('../../pav_newest_gladieux_only_fungap/sv_callers/del.',
                                          variable,'.txt', sep='')))$V1
  
  plot_df <- data.frame(
    value = c(vector_shuffled_rice, vector_del_rice),
    group = c(rep('MoO Baseline',length(vector_shuffled_rice)),
              rep('MoO Indels',length(vector_del_rice)))
  )
  
  p <- ggplot(data=plot_df, aes(x=value, y=group, fill=group))+
    geom_violin(lwd=0.25, scale="width", adjust=5) +
    geom_boxplot(width = 0.15, outlier.shape = NA,lwd=0.25,coef=0)+
          stat_summary(fun = "mean",
               geom = "point",
               color = "black",
               size=0.5)+
    xlab(legend_label)+
    ylab('')+
    xlim(axis_xlim_lower,axis_xlim_upper)+
    theme_classic()+
    theme(plot.title = element_text(size=8,hjust = 0.5), 
          axis.text=element_text(size=6),axis.title=element_text(size=8) , 
          legend.title=element_text(size=8), legend.position = 'bottom', 
          legend.text=element_text(size=6)) + theme(legend.title = 
                                                      element_blank())+
    theme(legend.position = "none")+
    scale_fill_manual(values=c(cbbPalette[4], "#5f5f5f", cbbPalette[2], cbbPalette[3]))
  
  ggsave(paste(variable, "_", 'density_plot',
               '.pdf',sep = ''), plot = p, width = 6.5/2, height = 7.5/4)

}

```