---
#MIT License
#
#Copyright (c) 2023 Pierre Michel Joubert
#
#Permission is hereby granted, free of charge, to any person obtaining a copy
#of this software and associated documentation files (the "Software"), to deal
#in the Software without restriction, including without limitation the rights
#to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#copies of the Software, and to permit persons to whom the Software is
#furnished to do so, subject to the following conditions:
#
#The above copyright notice and this permission notice shall be included in all
#copies or substantial portions of the Software.
#
#THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
#SOFTWARE.
title: "pav_distances"
output: html_document
date: '2022-10-18'
---

```{r}

library(ggplot2)
library(tidyr)
library(data.table)

```

```{r}


## read in gene file
read_locations = function(genome) {
  # list of elements to judge the distance of, in this case genes
  
  locations_file = paste("../random_forest/all_gffs_fixed/",genome,"_fungap_out.fixed.gff3", sep='')
  
  # read in file
  # just genes and their location and orientation
  locations <- read.table(locations_file,header=FALSE,sep='\t')
  
  locations <- locations[locations$V3 == 'gene',]
  
  locations <- subset(locations, select = c(V1, V4, V5, V7, V9))
  
  colnames(locations) <- c('scaffold', 'start', 'end', 'orientation', 'id')
  
  locations$id <- gsub(".*ID=(.+);.*", "\\1", locations$id)
  
  locations$id <- gsub("^(.+);.*", "\\1", locations$id)
  
  return(locations)
}

## read in te locations for comparison
read_comparison_te = function(genome) {
  comparison_file = paste("../random_forest/te_annotations/", genome,".w_class.gff", sep='')
    
  comparison <- read.table(comparison_file, header=FALSE,sep='\t')
  
  comparison <- subset(comparison, select = c(V1, V4, V5, V6, V7, V9, V10))
  
  colnames(comparison) <- c('scaffold', 'start', 'end', 'identity', 'orientation', 'element', 'class')
  
  return(comparison)
}

## read in gene locations for comparison
read_comparison_gene = function(genome) {
  comparison_file = paste("../random_forest/all_gffs_fixed/",genome,"_fungap_out.fixed.gff3", sep='')
    
  comparison <- read.table(comparison_file, header=FALSE,sep='\t')

  comparison <- comparison[comparison$V3 == 'gene',]
  
  comparison <- subset(comparison, select = c(V1, V4, V5, V7, V9))
  
  colnames(comparison) <- c('scaffold', 'start', 'end', 'orientation', 'id')
  
  comparison$id <- gsub(".*ID=(.+);.*", "\\1", comparison$id)
  
  comparison$id <- gsub("^(.+);.*", "\\1", locations$id)
  
  return(comparison)
}

# add orthogroup information
add_ogs = function(genome_full_name, locations) {
  
  orthogroup_info <- data.frame(df_ogs[genome_full_name])
  
  rownames(orthogroup_info) <- rownames(df_ogs)
  
  orthogroup_info <- na.omit(orthogroup_info)
  
  colnames(orthogroup_info) <- c('protein')
  
  ## this is to deal with the separation later, need to know how big to make the data frame
  max_columns <- max(sapply(orthogroup_info$protein, strsplit_sapply))
  
  column_names <- as.character(seq(from = 1, to = max_columns, by = 1))
  
  orthogroup_info <- separate(data = orthogroup_info, col = protein, into = column_names, sep = ", ")
  
  og_list <- list()
  
  for (col in colnames(orthogroup_info)){
    og_list <- c(og_list, setNames(rownames(orthogroup_info[! is.na(orthogroup_info[[col]]), ]),
                                   na.omit(orthogroup_info[[col]])))
  }
  
  og_list_ordered <- og_list[locations$id]
  
  # replace null values with NAs
  og_list_ordered[sapply(og_list_ordered, is.null)] <- NA
  
  return(as.vector(unlist(og_list_ordered)))
}

strsplit_sapply = function(gene_string){
  return(length(strsplit(gene_string, ', ')[[1]]))
}

## calculate distance to nearest element
element_distance = function(x, output, subset_per_scaffold) {
  scaffold = x['scaffold']
  start = as.numeric(x['start'])
  end = as.numeric(x['end'])
  enough_space_gene = x['enough_space_gene']
  id = as.character(x['id'])
  orientation = as.character(x['orientation'])
  
  nearby_upstream = subset_per_scaffold[subset_per_scaffold$end < start
                      & subset_per_scaffold$id != id,] ## make sure you dont match to the same gene each time
  
  if (dim(nearby_upstream)[1] != 0){ # if not empty
     upstream_distance = abs(max(nearby_upstream$end) - start)
     }else {
       upstream_distance = NA
     }
  
  nearby_downstream = subset_per_scaffold[subset_per_scaffold$start > end
                      & subset_per_scaffold$id != id,] ## make sure you dont match to the same gene each time
  
  if (dim(nearby_downstream)[1] != 0){ # if not empty
    downstream_distance = abs(min(nearby_downstream$start) - end)
    } else {
       downstream_distance = NA
    }
  
  if (orientation == "-") { ## these are swapped if the orientation is backwards
    downstream_distance_old = downstream_distance
    upstream_distance_old = upstream_distance
    upstream_distance = downstream_distance_old
    downstream_distance = upstream_distance_old
  }
  
  if (output == "downstream"){
    return(downstream_distance)
  } else if (output == "upstream") {
    return(upstream_distance)
  } else if (output == "closest") {
    if (is.na(downstream_distance) & is.na(upstream_distance)){
      return(NA)
    } else {
    return(min(downstream_distance,upstream_distance,na.rm=TRUE))
    }
  }
}

read_chromsizes = function(genome) {
  chromsizes_file = paste("../random_forest/chrom_sizes/",genome,".chromsizes", sep='')
  
  chromsizes <- read.table(chromsizes_file,header=FALSE,sep='\t')
  
  colnames(chromsizes) <- c('scaffold','size')
  
  return(chromsizes)
}

## check if there is enough space around the gene to call true/false nearby
enough_space = function(x, distance) {
  scaffold = x['scaffold']
  start = as.numeric(x['start'])
  end = as.numeric(x['end'])
  scaffold_size = chromsizes$size[chromsizes$scaffold==scaffold]
  if (start >= distance | end <= (scaffold_size - distance)){ ## check if there are enough sequenced base pairs to call a FALSE on at least one side
      return(TRUE)
  } else {
    return(FALSE)
  }
}

# call if a te is nearby true/false
nearby_te_per_class = function(x, class, distance) {
  scaffold = x['scaffold']
  start = as.numeric(x['start'])
  end = as.numeric(x['end'])
  enough_space_te = x['enough_space_te']
  nearby_te = subset_per_class_per_scaffold[subset_per_class_per_scaffold$end >= start - distance
                      & subset_per_class_per_scaffold$start <= end + distance,]
  if (dim(nearby_te)[1] == 0){ # if empty
    if (enough_space_te == TRUE){ ## check if there are enough sequenced base pairs to call a FALSE
      return(FALSE)
    } else {
      return(NA) ## otherwise we dont know
    }
  } else{
    return(TRUE)
  }
}

# call if a gene is nearby true/false
nearby_gene = function(x, class, distance) {
  scaffold = x['scaffold']
  start = as.numeric(x['start'])
  end = as.numeric(x['end'])
  enough_space_gene = x['enough_space_gene']
  id = x['id']
  nearby_genes = subset_genes_per_scaffold[subset_genes_per_scaffold$end >= start - distance
                      & subset_genes_per_scaffold$start <= end + distance
                      & subset_genes_per_scaffold$id != id,] ## make sure you dont match to the same gene each time
  if (dim(nearby_genes)[1] == 0){ # if empty
    if (enough_space_gene == TRUE){ ## check if there are enough sequenced base pairs to call a FALSE
      return(FALSE)
    } else {
      return(NA) ## otherwise we dont know
    }
  } else{
    return(TRUE)
  }
}

```

```{r}

## read in ogs
df_ogs <- data.frame(fread('../pipeline_methods/Orthogroups.tsv', na.strings = ''))

rownames(df_ogs) <- df_ogs$Orthogroup

df_ogs <- subset(df_ogs, select = -c(Orthogroup))

## change names to match lineage info spreadsheet

str_split_vector <- function(x, string){
  output_vector <- c()
  for ( item in x ) {
    output <- strsplit(item, string)[[1]][1]
    output_vector <- c(output_vector, output)
  }
  return(output_vector)
}


# fix col names
colnames(df_ogs) <- str_split_vector(colnames(df_ogs), '_')


colnames(df_ogs)[startsWith(colnames(df_ogs),'X')] <- gsub('X', '', colnames(df_ogs)[startsWith(colnames(df_ogs),'X')])


colnames(df_ogs) <- gsub('\\.', '-', colnames(df_ogs))

## read in lineage info
assembly_stats <- data.frame(fread('../pipeline_methods/gladieux_et_al_assembly_stats.txt', na.strings = ''))

lineage_info <- subset(assembly_stats, select = c(V1, V3, V4))

colnames(lineage_info) <- c('isolate', 'lineage', 'sublineage')

lineage_info[lineage_info$isolate == 'WD-3-1_1',]$isolate <- 'WD-3-1'

lineage_info$lineage_string <- paste(lineage_info$isolate, lineage_info$lineage, lineage_info$sublineage, rep("protein", length(lineage_info$sublineage)),sep = '_')
```

## TEST ##

```{r}

lineage <- 2

conserved_file <- paste('../pipeline_methods/conserved_ogs_', lineage, '.txt', sep = '')

conserved_ogs <- read.table(conserved_file, header=FALSE, sep='\t')

conserved_ogs <- conserved_ogs$V1

pav_file <- paste('../pipeline_methods/actual_pav_ogs_', lineage,'.txt', sep ='')

pav_ogs <- read.table(pav_file, header=FALSE, sep='\t')

pav_ogs <- pav_ogs$V1

genome <- "13FM-3-2"

genome_full_name <- "13FM-3-2"

```

```{r}

locations <- read_locations(genome)
comparison_tes <- read_comparison_te(genome)
comparison_genes <- read_comparison_gene(genome)

# add og information to locations AND comparison
locations$orthogroups <- suppressWarnings(add_ogs(genome_full_name, locations))
# drop genes without ogs
locations <- locations[! (is.na(locations$orthogroups)),]
comparison_genes$orthogroups <- suppressWarnings(add_ogs(genome_full_name, comparison_genes))

## make sure locations is sorted by scaffold

locations = locations[order(locations$scaffold),]

outputs = c("upstream", "downstream", "closest")

for (output in outputs){
  
  distance_column = c()
  
  for (scaffold in unique(locations$scaffold)){
    locations_per_scaffold = locations[locations$scaffold == scaffold,]
    subset_per_scaffold = comparison_genes[comparison_genes$scaffold == scaffold,]
    distance_column = c(distance_column, apply(locations_per_scaffold,1,element_distance,output=output, subset_per_scaffold=subset_per_scaffold))
  }
  
  locations[paste("gene_distance_", output, sep='')] = distance_column
  
}

class = "any_te"

if (class == "any_te"){
  subset_per_class = comparison_tes
} else {
  subset_per_class = comparison_tes[comparison_tes$class == class,]
}

for (output in outputs) {
  if (dim(subset_per_class)[1] != 0){
    class_column = c()
    for (scaffold in unique(locations$scaffold)){
      locations_per_scaffold = locations[locations$scaffold == scaffold,]
      subset_per_class_per_scaffold = subset_per_class[subset_per_class$scaffold == scaffold,]
      if (dim(subset_per_class_per_scaffold)[1] != 0){
        locations_per_scaffold[[class]] = apply(locations_per_scaffold,1,element_distance,output=output,
                                                subset_per_scaffold=subset_per_class_per_scaffold)
      } else {
        locations_per_scaffold[[class]] = NA ## if there are no TEs of that class in that genome then set all to NA
      }
      class_column = c(class_column, locations_per_scaffold[[class]])
    }
    locations[paste(class, "_distance_", output, sep='')] = class_column
  } else {
    locations[paste(class, "_distance_", output, sep='')] = NA ## if there are no TEs of that class in that genome then set all to NA
  }
}

```

```{r}
chromsizes <- read_chromsizes(genome_full_name)

distance_te = 5000

locations$enough_space_te = apply(locations,1,enough_space,distance=distance_te)

distance_gene = 1000

locations$enough_space_gene = apply(locations,1,enough_space,distance=distance_gene)

class = "any_te"

if (class == "any_te"){
  subset_per_class = comparison_tes
} else {
  subset_per_class = comparison_tes[comparison_tes$class == class,]
}
if (dim(subset_per_class)[1] != 0){
  class_column = c()
  for (scaffold in unique(locations$scaffold)){
    locations_per_scaffold = locations[locations$scaffold == scaffold,]
    subset_per_class_per_scaffold = subset_per_class[subset_per_class$scaffold == scaffold,]
    if (dim(subset_per_class_per_scaffold)[1] != 0){
      locations_per_scaffold[[class]] = apply(locations_per_scaffold,1,nearby_te_per_class,distance=distance_te)
    } else {
      locations_per_scaffold[[class]][locations_per_scaffold$enough_space_te == TRUE] = FALSE
    }
    class_column = c(class_column, locations_per_scaffold[[class]])
  }
  locations[[class]] = class_column
} else {
  locations[[class]][locations$enough_space_te == TRUE] = FALSE ## if there are no TEs of that class in that genome then set all to FALSE
}

gene_column = c()

for (scaffold in unique(locations$scaffold)){
  locations_per_scaffold = locations[locations$scaffold == scaffold,]
  subset_genes_per_scaffold = comparison_genes[comparison_genes$scaffold == scaffold,]
  # there will always be at least one gene...
  gene_column = c(gene_column, apply(locations_per_scaffold,1,nearby_gene,distance=distance_gene))
}

locations['gene_nearby'] = gene_column

```

## FULL RUN ###

```{r}

# pre-set up df to fill it with gene info later
df_distance_info <- data.frame(matrix(ncol = 20, nrow = 0))

colnames(df_distance_info) <- c("scaffold", "start", "end", "orientation", "id", "orthogroups", 
                            "gene_distance_upstream", "gene_distance_downstream", "gene_distance_closest",
                            "any_te_distance_upstream", "any_te_distance_downstream", "any_te_distance_closest",
                            "enough_space_te", "enough_space_gene",
                            "any_te", "gene_nearby",
                            "lineage_conserved", "lineage_pav",
                            "genome", "lineage"
                            )

for (lineage in c(2,3,4)) {
  genome_list <- lineage_info$isolate[lineage_info$lineage == lineage]

  # read in conserved ogs for lineage

  
  conserved_file <- paste('../pipeline_methods/conserved_ogs_', lineage, '.txt', sep = '')
  
  conserved_ogs <- read.table(conserved_file, header=FALSE, sep='\t')
  
  conserved_ogs <- conserved_ogs$V1
  
  # read in pav ogs for lineage
  
  pav_file <- paste('../pipeline_methods/actual_pav_ogs_', lineage,'.txt', sep ='')
  
  pav_ogs <- read.table(pav_file, header=FALSE, sep='\t')
  
  pav_ogs <- pav_ogs$V1
  
  for (index in seq(from = 1, to = length(genome_list))){
    
    genome <- genome_list[index]
    
    genome_full_name <- genome
    
    print(genome)
    # read in files for comparisons
    locations <- read_locations(genome)
    comparison_tes <- read_comparison_te(genome)
    comparison_genes <- read_comparison_gene(genome)
    
    # add og information to locations AND comparison
    locations$orthogroups <- suppressWarnings(add_ogs(genome_full_name, locations))
    # drop genes without ogs
    locations <- locations[! (is.na(locations$orthogroups)),]
    comparison_genes$orthogroups <- suppressWarnings(add_ogs(genome_full_name, comparison_genes))
    
    ## make sure locations is sorted by scaffold
    
    locations = locations[order(locations$scaffold),]
    
    outputs = c("upstream", "downstream", "closest")
    
    # calculate distance to nearest gene for all genes
    for (output in outputs){
      
      distance_column = c()
      
      for (scaffold in unique(locations$scaffold)){
        locations_per_scaffold = locations[locations$scaffold == scaffold,]
        subset_per_scaffold = comparison_genes[comparison_genes$scaffold == scaffold,]
        distance_column = c(distance_column, apply(locations_per_scaffold,1,element_distance,output=output,
                                                   subset_per_scaffold=subset_per_scaffold))
      }
      
      locations[paste("gene_distance_", output, sep='')] = distance_column
      
    }
    
    class = "any_te" ## only do all tes, not specific te classes
    
    if (class == "any_te"){
      subset_per_class = comparison_tes
    } else {
      subset_per_class = comparison_tes[comparison_tes$class == class,]
    }
    
    for (output in outputs) {
      if (dim(subset_per_class)[1] != 0){
        class_column = c()
        for (scaffold in unique(locations$scaffold)){
          locations_per_scaffold = locations[locations$scaffold == scaffold,]
          subset_per_class_per_scaffold = subset_per_class[subset_per_class$scaffold == scaffold,]
          if (dim(subset_per_class_per_scaffold)[1] != 0){
            locations_per_scaffold[[class]] = apply(locations_per_scaffold,1,element_distance,output=output,
                                                    subset_per_scaffold=subset_per_class_per_scaffold)
          } else {
            locations_per_scaffold[[class]] = NA ## if there are no TEs of that class in that genome then set all to NA
          }
          class_column = c(class_column, locations_per_scaffold[[class]])
        }
        locations[paste(class, "_distance_", output, sep='')] = class_column
      } else {
        locations[paste(class, "_distance_", output, sep='')] = NA ## if there are no TEs of that class in that genome then set all to NA
      }
    }
    
    # read in chromsizes file to get ending of scaffold locations (useful to check if enough space is there)
    chromsizes <- read_chromsizes(genome_full_name)
    # set minimum distance to call whether a te/gene gene is nearby
    distance_te = 5000
    
    locations$enough_space_te = apply(locations,1,enough_space,distance=distance_te)
    
    distance_gene = 1000
    
    locations$enough_space_gene = apply(locations,1,enough_space,distance=distance_gene)
    
    class = "any_te" ## only do all tes, not specific te classes
  
    ## loop per scaffold to speed things up
    # call true/false whether a te is nearby
    if (class == "any_te"){
      subset_per_class = comparison_tes
    } else {
      subset_per_class = comparison_tes[comparison_tes$class == class,]
    }
    if (dim(subset_per_class)[1] != 0){
      class_column = c()
      for (scaffold in unique(locations$scaffold)){
        locations_per_scaffold = locations[locations$scaffold == scaffold,]
        subset_per_class_per_scaffold = subset_per_class[subset_per_class$scaffold == scaffold,]
        if (dim(subset_per_class_per_scaffold)[1] != 0){
          locations_per_scaffold[[class]] = apply(locations_per_scaffold,1,nearby_te_per_class,distance=distance_te)
        } else {
          locations_per_scaffold[[class]][locations_per_scaffold$enough_space_te == TRUE] = FALSE
        }
        class_column = c(class_column, locations_per_scaffold[[class]])
      }
      locations[[class]] = class_column
    } else {
      locations[[class]][locations$enough_space_te == TRUE] = FALSE ## if there are no TEs of that class in that genome then set all to FALSE
    }
    
    gene_column = c()

    ## loop per scaffold to speed things up
    # call true/false whether a gene is nearby
    for (scaffold in unique(locations$scaffold)){
      locations_per_scaffold = locations[locations$scaffold == scaffold,]
      subset_genes_per_scaffold = comparison_genes[comparison_genes$scaffold == scaffold,]
      # there will always be at least one gene...
      gene_column = c(gene_column, apply(locations_per_scaffold,1,nearby_gene,distance=distance_gene))
    }
    
    locations['gene_nearby'] = gene_column
    
    locations$lineage_conserved <- locations$orthogroups %in% conserved_ogs
    
    locations$lineage_pav <- locations$orthogroups %in% pav_ogs
    
    locations$genome <- genome
    
    locations$lineage <- lineage
    
    # reorder before merging with the rest
    locations <- locations[,colnames(df_distance_info)]

    df_distance_info <- rbind(df_distance_info, locations)

    
  }
  
}

```

```{r}

output.file <- file('gene_distance_info.txt', "wb") ## to get unix line endings

write.table(df_distance_info, file = output.file, quote = FALSE, col.names = FALSE, row.names = FALSE)

close(output.file)

```
